### PR TITLE
Vendor cfssl/cfssljson utilities

### LINKED ISSUE #60070: local-up-cluster: cfssl is downloaded everytime, it should be downloaded just once
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

If you don't have cfssl in your `PATH`, then `local-up-cluster` script is downloading it on every execution, instead of downloading it once. It increases the execution time.

**What you expected to happen**:

If you don't have cfssl in `PATH`, `local-up-cluster.sh` script should download/install it only once and the next execution of script should just use the installed cfssl.

**How to reproduce it (as minimally and precisely as possible)**:

1. Ensure that you don't have `cfssl`
2. Execute `local-up-cluster.sh` script multiple times and observe that `cfssl` is downloaded every time.

**Anything else we need to know?**:

I think that the approach in the following function:

https://github.com/kubernetes/kubernetes/blob/master/hack/lib/util.sh#L740-L755

does not work and is too complicated. IMO we could just do `go get -u github.com/cloudflare/cfssl/cmd/...` if `cfssl` is not present in `PATH` and that's all. That script is for developers and it expects to have a working Golang environment, so I don't see any point in fetching OS-specific binary.

**Environment**:
- Kubernetes version (use `kubectl version`): master
- Cloud provider or hardware configuration: doesn't matter
- OS (e.g. from /etc/os-release):doesn't matter
- Kernel (e.g. `uname -a`): doesn't matter
- Install tools: local-up-cluster.sh
- Others:

### LINKED ISSUE #66995: downloaded CFSSL is broken on mac
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**: cfssl fails to work correctly when e2e testing from macOS

**What you expected to happen**: e2e testing with `go run hack/e2e.go -- --up ...` should bring up a cluster

**How to reproduce it (as minimally and precisely as possible)**: run `go run hack/e2e.go -- --build --up` on mac.

**Anything else we need to know?**: https://github.com/golang/go/issues/17335 looks related, possibly this binary is just built with too old of a go version?

**Environment**: macOS High Sierra 10.13.6

Relevant log snippet:
<details>

```console
+ kube::util::ensure-cfssl /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl
+ command -v cfssl
+ local cfssldir=/var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl
+ [[ -z /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl ]]
+ mkdir -p /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl
+ pushd /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl
+ command pushd /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl
+ pushd /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl
+ echo 'Unable to successfully run '\''cfssl'\'' from /Users/bentheelder/homebrew/opt/openssl/bin:/Users/bentheelder/Downloads/google-cloud-sdk/bin:/Users/bentheelder/.gvm/bin:/Users/bentheelder/homebrew/bin:/usr/local/git/current/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/local/go/bin:/Users/bentheelder/Downloads/google-cloud-sdk/bin:/Users/bentheelder/.gvm/bin:/Users/bentheelder/homebrew/bin:/Users/bentheelder/go/bin:/Users/bentheelder/go/bin; downloading instead...'
Unable to successfully run 'cfssl' from /Users/bentheelder/homebrew/opt/openssl/bin:/Users/bentheelder/Downloads/google-cloud-sdk/bin:/Users/bentheelder/.gvm/bin:/Users/bentheelder/homebrew/bin:/usr/local/git/current/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/local/go/bin:/Users/bentheelder/Downloads/google-cloud-sdk/bin:/Users/bentheelder/.gvm/bin:/Users/bentheelder/homebrew/bin:/Users/bentheelder/go/bin:/Users/bentheelder/go/bin; downloading instead...
++ uname -s
+ kernel=Darwin
+ case "${kernel}" in
+ curl --retry 10 -L -o cfssl https://pkg.cfssl.org/R1.2/cfssl_darwin-amd64
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  9.8M  100  9.8M    0     0  17.5M      0 --:--:-- --:--:-- --:--:-- 17.5M
+ curl --retry 10 -L -o cfssljson https://pkg.cfssl.org/R1.2/cfssljson_darwin-amd64
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 2176k  100 2176k    0     0  5020k      0 --:--:-- --:--:-- --:--:-- 5027k
+ chmod +x cfssl
+ chmod +x cfssljson
+ CFSSL_BIN=/var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl/cfssl
+ CFSSLJSON_BIN=/var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl/cfssljson
+ [[ ! -x /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl/cfssl ]]
+ [[ ! -x /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl/cfssljson ]]
+ popd
+ command popd
+ popd
+ echo '{"signing":{"default":{"expiry":"43800h","usages":["signing","key encipherment","client auth"]}}}'
+ echo '{"CN":"kubelet","names":[{"O":"system:nodes"}],"hosts":[""],"key":{"algo":"rsa","size":2048}}'
+ /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl/cfssl gencert -ca=pki/ca.crt -ca-key=pki/private/ca.key-config=ca-config.json -
+ /var/folders/hl/0xw19s6d3ln687jsdd243w8r00fsfv/T/tmp.gHWIiVrn/cfssl/cfssljson -bare kubelet
failed MSpanList_Insert 0x246000 0x5545d0ef825e 0x0 0x0
fatal error: MSpanList_Insert

runtime stack:
runtime.throw(0x15d320, 0x10)
        /usr/local/go/src/runtime/panic.go:530 +0x90 fp=0x7ffeefbfed50 sp=0x7ffeefbfed38
runtime.(*mSpanList).insert(0x1fdee8, 0x246000)
        /usr/local/go/src/runtime/mheap.go:933 +0x293 fp=0x7ffeefbfed80 sp=0x7ffeefbfed50
runtime.(*mheap).freeSpanLocked(0x1fd6e0, 0x246000, 0x100, 0x0)
        /usr/local/go/src/runtime/mheap.go:809 +0x4be fp=0x7ffeefbfede8 sp=0x7ffeefbfed80
runtime.(*mheap).grow(0x1fd6e0, 0x8, 0x0)
        /usr/local/go/src/runtime/mheap.go:675 +0x2a0 fp=0x7ffeefbfee40 sp=0x7ffeefbfede8
runtime.(*mheap).allocSpanLocked(0x1fd6e0, 0x1, 0x0)
        /usr/local/go/src/runtime/mheap.go:553 +0x4e3 fp=0x7ffeefbfee98 sp=0x7ffeefbfee40
runtime.(*mheap).alloc_m(0x1fd6e0, 0x1, 0x15, 0x0)
        /usr/local/go/src/runtime/mheap.go:437 +0x119 fp=0x7ffeefbfeec8 sp=0x7ffeefbfee98
runtime.(*mheap).alloc.func1()
        /usr/local/go/src/runtime/mheap.go:502 +0x41 fp=0x7ffeefbfeef8 sp=0x7ffeefbfeec8
runtime.systemstack(0x7ffeefbfef18)
        /usr/local/go/src/runtime/asm_amd64.s:307 +0xab fp=0x7ffeefbfef00 sp=0x7ffeefbfeef8
runtime.(*mheap).alloc(0x1fd6e0, 0x1, 0x10000000015, 0x108df)
        /usr/local/go/src/runtime/mheap.go:503 +0x63 fp=0x7ffeefbfef48 sp=0x7ffeefbfef00
runtime.(*mcentral).grow(0x1ff2e0, 0x0)
        /usr/local/go/src/runtime/mcentral.go:209 +0x93 fp=0x7ffeefbfefb0 sp=0x7ffeefbfef48
runtime.(*mcentral).cacheSpan(0x1ff2e0, 0x1fa308)
        /usr/local/go/src/runtime/mcentral.go:89 +0x47d fp=0x7ffeefbfeff0 sp=0x7ffeefbfefb0
runtime.(*mcache).refill(0x242000, 0x15, 0x7ffeefbff058)
        /usr/local/go/src/runtime/mcache.go:119 +0xcc fp=0x7ffeefbff028 sp=0x7ffeefbfeff0
runtime.mallocgc.func2()
        /usr/local/go/src/runtime/malloc.go:642 +0x2b fp=0x7ffeefbff048 sp=0x7ffeefbff028
runtime.systemstack(0x7ffeefbff0e8)
        /usr/local/go/src/runtime/asm_amd64.s:307 +0xab fp=0x7ffeefbff050 sp=0x7ffeefbff048
runtime.mallocgc(0x180, 0x140400, 0x0, 0x800000000)
        /usr/local/go/src/runtime/malloc.go:643 +0x869 fp=0x7ffeefbff128 sp=0x7ffeefbff050
runtime.newobject(0x140400, 0x1fa5b0)
        /usr/local/go/src/runtime/malloc.go:781 +0x42 fp=0x7ffeefbff150 sp=0x7ffeefbff128
runtime.malg(0x8000, 0x1fa820)
        /usr/local/go/src/runtime/proc.go:2634 +0x27 fp=0x7ffeefbff188 sp=0x7ffeefbff150
runtime.mpreinit(0x1fab60)
        /usr/local/go/src/runtime/os1_darwin.go:140 +0x1f fp=0x7ffeefbff1a0 sp=0x7ffeefbff188
runtime.mcommoninit(0x1fab60)
        /usr/local/go/src/runtime/proc.go:494 +0x105 fp=0x7ffeefbff1e8 sp=0x7ffeefbff1a0
runtime.schedinit()
        /usr/local/go/src/runtime/proc.go:434 +0x79 fp=0x7ffeefbff230 sp=0x7ffeefbff1e8
runtime.rt0_go(0x7ffeefbff268, 0x3, 0x7ffeefbff268, 0x0, 0x0, 0x3, 0x7ffeefbff4a0, 0x7ffeefbff4ee, 0x7ffeefbff4f4, 0x0, ...)
        /usr/local/go/src/runtime/asm_amd64.s:138 +0x132 fp=0x7ffeefbff238 sp=0x7ffeefbff230
failed MSpanList_Insert 0xa0f000 0x5545d283be65 0x0 0x0
fatal error: MSpanList_Insert

runtime stack:
runtime.throw(0x6bbbe0, 0x10)
        /usr/local/go/src/runtime/panic.go:530 +0x90 fp=0x7ffeefbfed00 sp=0x7ffeefbfece8
runtime.(*mSpanList).insert(0x9436e8, 0xa0f000)
        /usr/local/go/src/runtime/mheap.go:933 +0x293 fp=0x7ffeefbfed30 sp=0x7ffeefbfed00
runtime.(*mheap).freeSpanLocked(0x942ee0, 0xa0f000, 0x100, 0x0)
        /usr/local/go/src/runtime/mheap.go:809 +0x4be fp=0x7ffeefbfed98 sp=0x7ffeefbfed30
runtime.(*mheap).grow(0x942ee0, 0x8, 0x0)
        /usr/local/go/src/runtime/mheap.go:675 +0x2a0 fp=0x7ffeefbfedf0 sp=0x7ffeefbfed98
runtime.(*mheap).allocSpanLocked(0x942ee0, 0x1, 0x0)
        /usr/local/go/src/runtime/mheap.go:553 +0x4e3 fp=0x7ffeefbfee48 sp=0x7ffeefbfedf0
runtime.(*mheap).alloc_m(0x942ee0, 0x1, 0x15, 0x0)
        /usr/local/go/src/runtime/mheap.go:437 +0x119 fp=0x7ffeefbfee78 sp=0x7ffeefbfee48
runtime.(*mheap).alloc.func1()
        /usr/local/go/src/runtime/mheap.go:502 +0x41 fp=0x7ffeefbfeea8 sp=0x7ffeefbfee78
runtime.systemstack(0x7ffeefbfeec8)
        /usr/local/go/src/runtime/asm_amd64.s:307 +0xab fp=0x7ffeefbfeeb0 sp=0x7ffeefbfeea8
runtime.(*mheap).alloc(0x942ee0, 0x1, 0x10000000015, 0xee2f)
        /usr/local/go/src/runtime/mheap.go:503 +0x63 fp=0x7ffeefbfeef8 sp=0x7ffeefbfeeb0
runtime.(*mcentral).grow(0x944ae0, 0x0)
        /usr/local/go/src/runtime/mcentral.go:209 +0x93 fp=0x7ffeefbfef60 sp=0x7ffeefbfeef8
runtime.(*mcentral).cacheSpan(0x944ae0, 0x93d4e8)
        /usr/local/go/src/runtime/mcentral.go:89 +0x47d fp=0x7ffeefbfefa0 sp=0x7ffeefbfef60
runtime.(*mcache).refill(0xa0b000, 0x15, 0x7ffeefbff008)
        /usr/local/go/src/runtime/mcache.go:119 +0xcc fp=0x7ffeefbfefd8 sp=0x7ffeefbfefa0
runtime.mallocgc.func2()
        /usr/local/go/src/runtime/malloc.go:642 +0x2b fp=0x7ffeefbfeff8 sp=0x7ffeefbfefd8
runtime.systemstack(0x7ffeefbff098)
        /usr/local/go/src/runtime/asm_amd64.s:307 +0xab fp=0x7ffeefbff000 sp=0x7ffeefbfeff8
runtime.mallocgc(0x180, 0x6558a0, 0x0, 0x800000000)
        /usr/local/go/src/runtime/malloc.go:643 +0x869 fp=0x7ffeefbff0d8 sp=0x7ffeefbff000
runtime.newobject(0x6558a0, 0x93da10)
        /usr/local/go/src/runtime/malloc.go:781 +0x42 fp=0x7ffeefbff100 sp=0x7ffeefbff0d8
runtime.malg(0x8000, 0x93df00)
        /usr/local/go/src/runtime/proc.go:2634 +0x27 fp=0x7ffeefbff138 sp=0x7ffeefbff100
runtime.mpreinit(0x93e760)
        /usr/local/go/src/runtime/os1_darwin.go:140 +0x1f fp=0x7ffeefbff150 sp=0x7ffeefbff138
runtime.mcommoninit(0x93e760)
        /usr/local/go/src/runtime/proc.go:494 +0x105 fp=0x7ffeefbff198 sp=0x7ffeefbff150
runtime.schedinit()
        /usr/local/go/src/runtime/proc.go:434 +0x79 fp=0x7ffeefbff1e0 sp=0x7ffeefbff198
runtime.rt0_go(0x7ffeefbff218, 0x6, 0x7ffeefbff218, 0x0, 0x0, 0x6, 0x7ffeefbff468, 0x7ffeefbff4b2, 0x7ffeefbff4ba, 0x7ffeefbff4c9, ...)
        /usr/local/go/src/runtime/asm_amd64.s:138 +0x132 fp=0x7ffeefbff1e8 sp=0x7ffeefbff1e0
```

</details>

### PR DESCRIPTION
**Pull request by cblecker**
**What this PR does / why we need it**:
Vendors the `cfssl` and `cfssljson` tools. Updates `kube::util::ensure-cfssl` to use them.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
fixes #66995, fixes #60070

**Special notes for your reviewer**:
1. Add cfssl/cfssljson ot the required bins for saving
2. Manually cloned/checked out the new dependencies to my gopath. `godep restore` doesn't pull them down because they aren't required or already in the `Godeps.json`. Used @BenTheElder's list here: https://github.com/kubernetes/kubernetes/issues/66995#issuecomment-410594532
3. `hack/godep-save.sh` to add the packages and dependencies to godep
4. Fixed two bugs when building:
  a. `golang.org/x/crypto` needed to be updated
  b. `github.com/cloudflare/cfssl` needed to be updated to https://github.com/cloudflare/cfssl/commit/56268a613adfed278936377c18b1152d2c4ad5da so we can vendor their fork of `crypto/tls`, as we discard their modified vendored stdlib.
5. Update staging godeps
6. Update the `kube::util::ensure-cfssl` to install from vendor

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: hack/lib/util.sh [hunk 1]
```diff
@@ -707,50 +707,15 @@ function kube::util::join {
...
-    return 0
-  fi
-
-  # Create a temp dir for cfssl if no directory was given
-  local cfssldir=${1:-}
-  if [[ -z "${cfssldir}" ]]; then
-    kube::util::ensure-temp-dir
-    cfssldir="${KUBE_TEMP}/cfssl"
+  kube::log::status "Installing cfssl from vendor"
+  GOBIN="${KUBE_OUTPUT_BINPATH}" go install k8s.io/kubernetes/vendor/github.com/cloudflare/cfssl/cmd/cfssl
```
Comment (BenTheElder) [inline hack/lib/util.sh]: do we require to be in `$GOPATH/k8s.io/kubernetes` for these rules typically? I've had good success elsewhere doing ` go install ./vendor/...`

**Review by BenTheElder — APPROVED** [review]
one question, otherwise LGTM, thank you!! 😄

File: hack/lib/util.sh (hunk 1 above)
Comment (cblecker) [reply hack/lib/util.sh]: Yes, we should be in GOPATH. Most operations rely on it, and it's how we're doing other utilities like kazel/gazelle/godep. Also easier than changing cwd to ensure we're in the right relative path.

Comment (BenTheElder) [reply hack/lib/util.sh]: SGTM. I've frequently heard complaints about kazel and gopath and wasn't sure if we more broadly have decided to require this. 👍
In the other places it's not a bash library so cd is fine 😛

**Comment by BenTheElder** [comment]
/lgtm
/hold
Holding so someone else can approve the CFSSL bump, overall this is _great_ but I'll let someone else verify that commit.

File: hack/lib/util.sh [hunk 2]
```diff
@@ -707,50 +707,15 @@ function kube::util::join {
...
-    CFSSLJSON_BIN=$(command -v cfssljson)
-    return 0
-  fi
-
-  # Create a temp dir for cfssl if no directory was given
-  local cfssldir=${1:-}
-  if [[ -z "${cfssldir}" ]]; then
-    kube::util::ensure-temp-dir
-    cfssldir="${KUBE_TEMP}/cfssl"
+  kube::log::status "Installing cfssl from vendor"
```
Comment (BenTheElder) [inline hack/lib/util.sh]: hmm 
```
W0809 04:55:27.758] /go/src/k8s.io/kubernetes/kubernetes/hack/lib/util.sh: line 710: kube::log::status: command not found
W0809 04:55:27.758] /go/src/k8s.io/kubernetes/kubernetes/hack/lib/util.sh: line 711: KUBE_OUTPUT_BINPATH: unbound variable
```

**Review by BenTheElder — APPROVED** [review]
/lgtm

**Review by BenTheElder — APPROVED** [review]
/lgtm

**Comment by cblecker** [comment]
Licence details:
A couple new deps of deps, but they are under either MIT or Mozilla (compatible). New packages under cfssl and golang/x/crypto (already vendored).

**Comment by cblecker** [comment]
/hold cancel
:shipit: 

@BenTheElder Added notes to PR body of how to generate.

**Comment by BenTheElder** [comment]
Thanks @cblecker ! :smile:

**Pull request merged**